csp/impl/types/pydantic_types.py

```diff
@@ -0,0 +1,223 @@
+import collections.abc
+import platform
+import sys
+import types
+import typing
+import typing_extensions
+from packaging import version
+from pydantic import GetCoreSchemaHandler, ValidationInfo, ValidatorFunctionWrapHandler
+from pydantic_core import CoreSchema, core_schema
+from typing import Any, ForwardRef, Generic, Optional, Type, TypeVar, Union, get_args, get_origin
+
+from csp.impl.types.common_definitions import OutputBasket, OutputBasketContainer
+from csp.impl.types.tstype import SnapKeyType, SnapType, isTsDynamicBasket
+from csp.impl.types.typing_utils import TsTypeValidator
+
+# Required for py38 compatibility
+# In python 3.8, get_origin(List[float]) returns list, but you can't call list[float] to retrieve the annotation
+# Furthermore, Annotated is part of typing_Extensions and get_origin(Annotated[str, ...]) returns str rather than Annotated
+_IS_PY38 = version.parse(platform.python_version()) < version.parse("3.9")
+# For a more complete list, see https://github.com/alexmojaki/eval_type_backport/blob/main/eval_type_backport/eval_type_backport.py
+_PY38_ORIGIN_MAP = {
+    tuple: typing.Tuple,
+    list: typing.List,
+    dict: typing.Dict,
+    set: typing.Set,
+    frozenset: typing.FrozenSet,
+    collections.abc.Callable: typing.Callable,
+    collections.abc.Iterable: typing.Iterable,
+    collections.abc.Mapping: typing.Mapping,
+    collections.abc.MutableMapping: typing.MutableMapping,
+    collections.abc.Sequence: typing.Sequence,
+}
+
+_K = TypeVar("T", covariant=True)
+_T = TypeVar("T", covariant=True)
+
+
+def _check_source_type(cls, source_type):
+    """Helper function for CspTypeVarType and CspTypeVar"""
+    args = get_args(source_type)
+    if len(args) != 1:
+        raise ValueError(f"Must pass a single generic argument to {cls.__name__}. Got {args}.")
+    v = args[0]
+    if type(v) is TypeVar:
+        return v.__name__
+    elif type(v) is ForwardRef:  # In case someone writes, i.e. CspTypeVar["T"]
+        return v.__forward_arg__
+    else:
+        raise ValueError(f"Must pass either a TypeVar or a ForwardRef (string) to {cls.__name__}. Got {type(v)}.")
+
+
+class CspTypeVarType(Generic[_T]):
+    """A special type representing a template variable for csp.
+    It behaves similarly to a ForwardRef, but where the type of the forward arg is *implied* by the input type.
+    """
+
+    @classmethod
+    def __get_pydantic_core_schema__(cls, source_type: Any, handler: GetCoreSchemaHandler) -> CoreSchema:
+        typ = _check_source_type(cls, source_type)
+
+        def _validator(v: Any, info: ValidationInfo) -> Any:
+            # info.context should be an instance of TVarValidationContext, but we don't check for performance
+            if info.context is None:
```

Comment:
Technically it's all designed to work properly even if it wasn't all internal and this make sure that any errors that would arise have sensible messages (rather than "'NoneType' object has no attribute 'add_tvar_type_ref'").